alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
+
+        control.identifier = NSUserInterfaceItemIdentifier(rawName)
+        control.target = self
+        control.action = #selector(controlDidEndEditing)
+        let containerView = NSStackView(views: [label, control])
+
+        if suffixText != nil {
+            let suffix = NSTextField(labelWithString: suffixText!)
+            suffix.textColor = .gray
+            containerView.addView(suffix, in: .leading)
         }
+
+        return containerView
     }
 
-    func textDidEndEditing(_ notification: Notification) {
-        if let textView = notification.object as? NSTextView {
-            let key = inputsMap[textView]!
+    /*
+    usage notes:
+    - NSSlider: supports on purpose currently only decimal values
+    */
+    @objc func controlDidEndEditing(senderControl: NSControl) {
+        self.makeFirstResponder(invisibleTextField) // deselects any possibly selected NSTextField (so slider & popUp changes deselect them)
+
+        let key: String? = senderControl.identifier?.rawValue
+        var newValue: String?
+
+        if senderControl is NSPopUpButton {
+            newValue = (senderControl as! NSPopUpButton).titleOfSelectedItem!
+        } else if senderControl is NSSlider {
+            newValue = String(format: "%.0f", Double(senderControl.stringValue)!) // we are only interested in decimals of the provided double
+        } else {
+            newValue = senderControl.stringValue
+        }
+
+//        debugPrint("PreferencesPanel: save: change", key!, newValue!)
+
+        if key != nil && newValue != nil {
+            if newValue == Preferences.rawValues[key!] {
+                debugPrint("PreferencesPanel: save: abort: value was not changed")
+                return
+            }
+
+            let previousValue = Preferences.rawValues[key!]!
+
             do {
-                try Preferences.updateAndValidateFromString(key, textView.string)
+                try Preferences.updateAndValidateFromString(key!, newValue!)
                 try Preferences.saveRawToDisk()
             } catch {
-                debugPrint(key, error)
-                textView.string = Preferences.rawValues[key]!
+                debugPrint("PreferencesPanel: save: error", key!, error, "previousValue", previousValue, " | newValue", newValue!)
+
+                // restores the previous value in Preferences and senderControl
+                try! Preferences.updateAndValidateFromString(key!, previousValue)
+
+                if senderControl is NSPopUpButton {
+                    (senderControl as! NSPopUpButton).selectItem(withTitle: previousValue)
+                } else {
+                    senderControl.stringValue = previousValue
+                }
             }
+        } else {
+            debugPrint("PreferencesPanel: save: error: key||newValue = nil", key!, newValue!)
         }
     }
+
+    private func calcLabelWidth() -> CGFloat {
+        return (panelWidth - panelPadding) * panelWidthToLabelRatio
+    }
```

Comment:
These 3 values are static I think, so you can have this derived value as an attribute in the class instead of a dynamic method. You can either have it after the 3 attributes as a simple attribute, or use swift computed attribute language feature.

Reply:
I wanted to have this derived convenience value somewhere and initially had it setup as a class variable which then I had to set up in the init() of the class and it felt wrong for such a simple thing. Thats why I switched to a helper method for that value. Thanks for the hint about that language feature! 😎